nat/behavior/discovery.go

```diff
@@ -0,0 +1,372 @@
+	request.Add(stun.AttrChangeRequest, changeRequestAddressPort)
+
+	ctx1, cl = context.WithTimeout(ctx, timeout)
+	defer cl()
+	_, err = mapTestConn.roundTrip(ctx1, request, mapTestConn.RemoteAddr)
+	switch {
+	case err == nil:
+		return FilteringIndependent, nil
+	case err == ctx1.Err() && ctx.Err() == nil:
+		// Nothing, just no response. Proceed to next test.
+	default:
+		return "", fmt.Errorf("filtering test II failed: %w", err)
+	}
+
+	// Test III: Request to change port only
+	request = stun.MustBuild(stun.TransactionID, stun.BindingRequest)
+	request.Add(stun.AttrChangeRequest, changeRequestPort)
+
+	ctx1, cl = context.WithTimeout(ctx, timeout)
+	defer cl()
+	_, err = mapTestConn.roundTrip(ctx1, request, mapTestConn.RemoteAddr)
+	switch {
+	case err == nil:
+		return FilteringAddress, nil
+	case err == ctx1.Err() && ctx.Err() == nil:
+		return FilteringAddressPort, nil
+	default:
+		return "", fmt.Errorf("filtering test III failed: %w", err)
+	}
+}
+
+// Parse a STUN message
+func parse(msg *stun.Message) (ret struct {
+	xorAddr    *stun.XORMappedAddress
+	otherAddr  *stun.OtherAddress
+	mappedAddr *stun.MappedAddress
+	software   *stun.Software
+}) {
+	ret.mappedAddr = &stun.MappedAddress{}
+	ret.xorAddr = &stun.XORMappedAddress{}
+	ret.otherAddr = &stun.OtherAddress{}
+	ret.software = &stun.Software{}
+	if ret.xorAddr.GetFrom(msg) != nil {
+		ret.xorAddr = nil
+	}
+	if ret.otherAddr.GetFrom(msg) != nil {
+		ret.otherAddr = nil
+	}
+	if ret.mappedAddr.GetFrom(msg) != nil {
+		ret.mappedAddr = nil
+	}
+	if ret.software.GetFrom(msg) != nil {
+		ret.software = nil
+	}
+	return ret
+}
+
+// Given an address string, returns a StunServerConn
+func connect(address string) (*stunServerConn, error) {
+	addr, err := net.ResolveUDPAddr("udp4", address)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := net.ListenUDP("udp4", nil)
+	if err != nil {
+		return nil, err
+	}
+	mChan := listen(c)
+
+	return &stunServerConn{
+		conn:        c,
+		LocalAddr:   c.LocalAddr(),
+		RemoteAddr:  addr,
+		messageChan: mChan,
+	}, nil
+}
+
+// Send request and wait for response or timeout
+func (c *stunServerConn) roundTrip(ctx context.Context, msg *stun.Message, addr net.Addr) (*stun.Message, error) {
+	err := msg.NewTransactionID()
+	if err != nil {
+		return nil, err
+	}
+	_, err = c.conn.WriteTo(msg.Raw, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wait for response or timeout
+	for {
+		select {
+		case m, ok := <-c.messageChan:
+			if !ok {
+				return nil, ErrResponseMessage
+			}
+			if m.TransactionID == msg.TransactionID {
+				return m, nil
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
+// taken from https://github.com/pion/stun/blob/master/cmd/stun-traversal/main.go
+func listen(conn *net.UDPConn) (messages chan *stun.Message) {
+	messages = make(chan *stun.Message)
+	go func() {
+		for {
+			buf := make([]byte, 1024)
+
+			n, _, err := conn.ReadFromUDP(buf)
+			if err != nil {
+				close(messages)
```

Comment:
I think we could just call `defer close(messages)` at the top of `go func()`.

Reply:
Sure! Updated.